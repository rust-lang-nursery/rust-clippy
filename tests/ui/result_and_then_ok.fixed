// run-rustfix
#![deny(clippy::result_and_then_ok)]

pub fn main() {
    let x: Result<u32, &str> = Ok(5);
    // the easiest cases
    let _ = x;
    let _ = x.map(|o| o + 1);
    // and an easy counter-example
    let _ = x.and_then(|o| if o < 32 { Ok(o) } else { Err("error") });
}

pub fn foo() -> Result<String, ()> {
    let x = Ok(String::from("hello"));
    Ok("hello".to_owned()).and_then(|s| Ok(format!("{}{}", s, x?)))
}

pub fn example2(x: bool) -> Result<&'static str, ()> {
    Ok("a").and_then(|s| Ok(if x { s } else { return Err(()) }))
}
