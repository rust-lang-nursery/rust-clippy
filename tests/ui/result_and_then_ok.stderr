error: using `Result.and_then(Ok)`, which is a no-op
  --> $DIR/result_and_then_ok.rs:7:13
   |
LL |     let _ = x.and_then(Ok);
   |             ^^^^^^^^^^^^^^ help: use the expression directly: `x`
   |
note: lint level defined here
  --> $DIR/result_and_then_ok.rs:2:9
   |
LL | #![deny(clippy::result_and_then_ok)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^

error: using `Result.and_then(|x| Ok(y))`, which is more succinctly expressed as `map(|x| y)`
  --> $DIR/result_and_then_ok.rs:8:13
   |
LL |     let _ = x.and_then(|o| Ok(o + 1));
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `x.map(|o| o + 1)`

error: aborting due to 2 previous errors

